network/ocsp.go

```diff
@@ -0,0 +1,273 @@
+	}
+
+	switch required {
+	case "yes", "true":
+		log.Debugf("OCSP: At least one OCSP server should confirm certificate validity")
+	case "no", "false":
+		log.Debugf("OCSP: Allowing certificates not known by OCSP server")
+	case "all":
+		log.Debugf("OCSP: Requiring positive response from all OCSP servers")
+	}
+
+	switch fromCert {
+	case "use":
+		log.Debugf("OCSP: using servers described in certificates if nothing passed via command line")
+	case "trust":
+		log.Debugf("OCSP: trusting responses from OCSP servers listed in certificates")
+	case "prefer":
+		log.Debugf("OCSP: server from certificate will be prioritized over one from command line")
+	case "ignore":
+		log.Debugf("OCSP: ignoring OCSP servers described in certificates")
+	}
+
+	return &OCSPConfig{url: uri, required: requiredVal, fromCert: fromCertVal}, nil
+}
+
+// OCSPClient is used to perform OCSP queries to some URI
+type OCSPClient interface {
+	// Query generates OCSP request about specified certificate, sends it to server and returns the response
+	Query(commonName string, clientCert, issuerCert *x509.Certificate, ocspServerURL string) (*ocsp.Response, error)
+}
+
+// DefaultOCSPClient is a default implementation of OCSPClient
+type DefaultOCSPClient struct{}
+
+// Query generates OCSP request about specified certificate, sends it to server and returns the response
+func (c DefaultOCSPClient) Query(commonName string, clientCert, issuerCert *x509.Certificate, ocspServerURL string) (*ocsp.Response, error) {
+	opts := &ocsp.RequestOptions{Hash: crypto.SHA256}
+	buffer, err := ocsp.CreateRequest(clientCert, issuerCert, opts)
+	if err != nil {
+		return nil, err
+	}
+	httpRequest, err := http.NewRequest(http.MethodPost, ocspServerURL, bytes.NewBuffer(buffer))
+	if err != nil {
+		return nil, err
+	}
+	ocspURL, err := url.Parse(ocspServerURL)
+	if err != nil {
+		return nil, err
+	}
+	httpRequest.Header.Add("Content-Type", "application/ocsp-request")
+	httpRequest.Header.Add("Accept", "application/ocsp-response")
+	httpRequest.Header.Add("host", ocspURL.Host)
+	httpClient := &http.Client{}
+	httpResponse, err := httpClient.Do(httpRequest)
+	if err != nil {
+		return nil, err
+	}
+	defer httpResponse.Body.Close()
+	output, err := ioutil.ReadAll(httpResponse.Body)
+	if err != nil {
+		return nil, err
+	}
+	ocspResponse, err := ocsp.ParseResponse(output, issuerCert)
+	return ocspResponse, err
+}
+
+// DefaultOCSPVerifier is a default OCSP verifier
+type DefaultOCSPVerifier struct {
+	Config OCSPConfig
+	Client OCSPClient
+}
+
+// ocspServerToCheck is used to plan OCSP requests
+type ocspServerToCheck struct {
+	url      string
+	fromCert bool
+}
+
+// Verify ensures certificate is not revoked by querying configured OCSP servers
+func (v DefaultOCSPVerifier) Verify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	for _, chain := range verifiedChains {
+		log.Debugf("OCSP: Verifying '%s'", chain[0].Subject.String())
+
+		cert := chain[0]
```

Comment:
we should validate all input from unknown sources. We don't sure that there are at least 2 values in array.